### PR TITLE
Ignore qualifiers when doing autoimport completions lookup

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
A follow-up of https://github.com/rust-analyzer/rust-analyzer/pull/6918#issuecomment-748511151 and the PR itself.

Tweaks the `import_map` query api to be more flexible with the ways to match against the import path and now fuzzy imports search in names only.
This had improved the completion speed for me locally in ~5 times for `fuzzy_completion` span time, but please recheck me here.

IMO we're fast and presice enough now, so I've added the modules back to the fuzzy search output.

Also tweaks the the expect tests to display functions explicitly, to avoid confusing "duplicate" results.

### CODE REVIEW

File: crates/hir_def/src/import_map.rs [hunk 1]
```diff
@@ -407,14 +451,31 @@ mod tests {
...
                     let map = db.import_map(krate);
-                    let path = map.path_of(item).unwrap();
+
+                    let path = match assoc_to_trait(&db, item) {
+                        Some(trait_) => {
+                            let mut full_path = map.path_of(trait_).unwrap().to_string();
+                            if let ItemInNs::Types(ModuleDefId::FunctionId(function_id))
+                            | ItemInNs::Values(ModuleDefId::FunctionId(function_id)) = item
+                            {
+                                full_path += &format!(
```
Comment (lnicola) [inline crates/hir_def/src/import_map.rs:469]: Nit: `.push_str`

Comment (lnicola) [reply crates/hir_def/src/import_map.rs:469]: Ah, `Name` is harder to format. We could still use `write!`, I guess, but.. yeah.

**Comment by lnicola** [comment]
Well, I'm not completely certain that this is correct, but it's too good to not have it in nightly versions.

@matklad you should probably still review this when you're back.

File: crates/hir_def/src/import_map.rs (hunk 1 above)
Comment (matklad) [reply crates/hir_def/src/import_map.rs:469]: https://github.com/rust-analyzer/rust-analyzer/pull/7092

**Comment by matklad** [comment]
lgtm!

>you should probably still review this when you're back.

Yeah, I've decided to use the Christmass break to switch to the workman layout: https://workmanlayout.org. My productivity tanked,  but I've already clawed back like 20wpm! :D

**Comment by lnicola** [comment]
*shudders*

I tried Colemak for a while in another life, but gave up.

**Comment by Mirko-von-Leipzig** [comment]
> 
> 
> lgtm!
> 
> > you should probably still review this when you're back.
> 
> Yeah, I've decided to use the Christmass break to switch to the workman layout: https://workmanlayout.org. My productivity tanked, but I've already clawed back like 20wpm! :D

Would you recommend it over the old favourites (Colemak and Dvorak)?

**Comment by kjeremy** [comment]
Yeah we'll need a blog post about this.

**Comment by matklad** [comment]
I'll do one in about a year. 

From my research, dvorak seems significantly worse than modern alternatives (its from 30s!), and colemac, workman & noreman are roughly equivalent.

**Comment by matklad** [comment]
Also (and the thing that motivated the switch for me), Russian йцукен is believed to be better than qwerty, and I do type significantly faster in Russian, despite typing in English more overall. Not sure if this is layout or native vs foreign language thing.

**Comment by flodiebold** [comment]
FWIW, I'm using workman, and am mostly happy with it, but sometimes wish I'd have chosen a layout that's more widely supported ([see also](https://github.com/flodiebold/dotfiles/blob/master/emacs/.emacs.d/private/workman/packages.el)), and think that probably the differences between these layouts (colemak etc., not qwerty) matter much less than the effect of learning to touch-type properly.

**Comment by YaLTeR** [comment]
I clicked on a link in the changelog to check out an interesting change but found an interesting conversation. :)

> Also (and the thing that motivated the switch for me), Russian йцукен is believed to be better than qwerty, and I do type significantly faster in Russian, despite typing in English more overall. Not sure if this is layout or native vs foreign language thing.

I also type more in English (qwerty) than in Russian and I do type slower in Russian than in English so ¯\\\_(ツ)_/¯
